app/upgrades.go

```diff
@@ -1,3 +1,274 @@
+		})
+
+		// Halt iteration if there was an error
+		if iterErr != nil {
+			return true
+		}
+
+		// Aggregate sum of all fractional balances
+		aggregateSum = aggregateSum.Add(acc.Balance)
+
+		// Continue iterating
+		return false
+	})
+
+	return aggregateSum, iterErr
+}
+
+// InitializeRemainder initializes the remainder amount in x/precisebank. It
+// calculates the remainder amount that is needed to ensure that the sum of all
+// fractional balances is a multiple of the conversion factor. The remainder
+// amount is stored in the store and returned.
+func InitializeRemainder(
+	ctx sdk.Context,
+	precisebankKeeper precisebankkeeper.Keeper,
+	aggregateSum sdkmath.Int,
+) sdkmath.Int {
+	// Extra fractional coins that exceed the conversion factor.
+	// This extra + remainder should equal the conversion factor to ensure
+	// (sum(fBalances) + remainder) % conversionFactor = 0
+	extraFractionalAmount := aggregateSum.Mod(precisebanktypes.ConversionFactor())
+	remainder := precisebanktypes.ConversionFactor().
+		Sub(extraFractionalAmount).
+		// Mod conversion factor to ensure remainder is valid.
+		// If extraFractionalAmount is a multiple of conversion factor, the
+		// remainder is 0.
+		Mod(precisebanktypes.ConversionFactor())
+
+	// Panics if the remainder is invalid. In a correct chain state and only
+	// mint/burns due to transfers, this would be 0.
+	precisebankKeeper.SetRemainderAmount(ctx, remainder)
+
+	return remainder
+}
+
+// TransferFractionalBalanceReserve migrates the fractional balance reserve from
+// x/evmutil to x/precisebank. It transfers the reserve balance from x/evmutil
+// to x/precisebank and ensures that the reserve fully backs all fractional
+// balances. It mints or burns coins to back the fractional balances exactly.
+func TransferFractionalBalanceReserve(
+	ctx sdk.Context,
+	accountKeeper evmutiltypes.AccountKeeper,
+	bankKeeper bankkeeper.Keeper,
+	precisebankKeeper precisebankkeeper.Keeper,
+) error {
+	logger := ctx.Logger()
+
+	// Transfer x/evmutil reserve to x/precisebank.
+	evmutilAddr := accountKeeper.GetModuleAddress(evmutiltypes.ModuleName)
+	reserveBalance := bankKeeper.GetBalance(ctx, evmutilAddr, precisebanktypes.IntegerCoinDenom)
+
+	if err := bankKeeper.SendCoinsFromModuleToModule(
+		ctx,
+		evmutiltypes.ModuleName,     // from x/evmutil
+		precisebanktypes.ModuleName, // to x/precisebank
+		sdk.NewCoins(reserveBalance),
+	); err != nil {
+		return fmt.Errorf("failed to transfer reserve from x/evmutil to x/precisebank: %w", err)
+	}
+
+	logger.Info(fmt.Sprintf("transferred reserve balance: %s", reserveBalance))
+
+	// Ensure x/precisebank reserve fully backs all fractional balances.
+	totalFractionalBalances := precisebankKeeper.GetTotalSumFractionalBalances(ctx)
+
+	// Does NOT ensure state is correct, total fractional balances should be a
+	// multiple of conversion factor but is not guaranteed due to the remainder.
+	// Remainder initialization is handled by InitializeRemainder.
```

Comment:
similar to the above, because our upgrade handlers are specific to mainnet and testnet and we know there should be no remainder, maybe we should be making the stricter assertion here?

curious what @nddeluca thinks

Reply:
thanks for creating the tracking ticket. we will re-assess after the mainner mirror test 👍